sensing/autoware_calibration_status/lib/preprocess/preprocess_cuda.cu

```diff
@@ -0,0 +1,379 @@
+    const double & fy_new = projection_matrix[5];  // P[1,1]
+    const double & cx_new = projection_matrix[2];  // P[0,2]
+    const double & cy_new = projection_matrix[6];  // P[1,2]
+
+    const double x_proj = (x - cx_new) / fx_new;
+    const double y_proj = (y - cy_new) / fy_new;
+
+    // Apply forward distortion model
+    const double & k1 = dist_coeffs[0];
+    const double & k2 = dist_coeffs[1];
+    const double & p1 = dist_coeffs[2];
+    const double & p2 = dist_coeffs[3];
+    const double & k3 = dist_coeffs[4];
+    const double & k4 = dist_coeffs[5];
+    const double & k5 = dist_coeffs[6];
+    const double & k6 = dist_coeffs[7];
+
+    const double r2 = x_proj * x_proj + y_proj * y_proj;
+    const double r4 = r2 * r2;
+    const double r6 = r4 * r2;
+
+    const double radial_num = 1.0 + k1 * r2 + k2 * r4 + k3 * r6;
+    const double radial_den = 1.0 + k4 * r2 + k5 * r4 + k6 * r6;
+    const double radial_dist = radial_num / radial_den;
+
+    const double dx = 2.0 * p1 * x_proj * y_proj + p2 * (r2 + 2.0 * x_proj * x_proj);
+    const double dy = p1 * (r2 + 2.0 * y_proj * y_proj) + 2.0 * p2 * x_proj * y_proj;
+
+    const double x_distorted = x_proj * radial_dist + dx;
+    const double y_distorted = y_proj * radial_dist + dy;
+
+    // Project distorted point to source image pixel coordinates
+    const double & fx_orig = camera_matrix[0];  // K[0,0]
+    const double & fy_orig = camera_matrix[4];  // K[1,1]
+    const double & cx_orig = camera_matrix[2];  // K[0,2]
+    const double & cy_orig = camera_matrix[5];  // K[1,2]
+
+    const double u = fx_orig * x_distorted + cx_orig;
+    const double v = fy_orig * y_distorted + cy_orig;
+
+    // Sample source image and write to output
+    InputImageBGR8Type px;
+    if (u >= 0 && v >= 0 && u < width && v < height) {
+      px = bilinear_sample(input_image, u, v, width, height);
+    } else {
+      px = {0, 0, 0};
+    }
+
+    output_image[y * width + x] = px;
+    output_array[0 * height * width + y * width + x] = static_cast<float>(px.r) / 255.0f;
+    output_array[1 * height * width + y * width + x] = static_cast<float>(px.g) / 255.0f;
+    output_array[2 * height * width + y * width + x] = static_cast<float>(px.b) / 255.0f;
+    output_array[3 * height * width + y * width + x] = 0.0f;
+    output_array[4 * height * width + y * width + x] = 0.0f;
+  }
+}
+
+cudaError_t PreprocessCuda::undistortImage_launch(
+  const InputImageBGR8Type * input_image, const double * dist_coeffs, const double * camera_matrix,
+  const double * projection_matrix, const size_t width, const size_t height,
+  InputImageBGR8Type * output_image, float * output_array)
+{
+  dim3 threads(16, 16);
+  dim3 blocks((width + threads.x - 1) / threads.x, (height + threads.y - 1) / threads.y);
+
+  undistortImage_kernel<<<blocks, threads, 0, stream_>>>(
+    input_image, dist_coeffs, camera_matrix, projection_matrix, width, height, output_image,
+    output_array);
```

Comment:
This kernel call may be replaced by some NVIDIA-provided API, which runs fast in most cases and is flexible for various inputs, in particular, the camera distortion model
- zero-filling all elements in the output buffer
- [`nppiRemap_8u_C3R`](https://docs.nvidia.com/cuda/archive/9.2/npp/group__image__remap.html#ga62d5fb5fcd3e3bc67e06830c41ad339d) for undistortion
- [`nppiCopy_8u_C3P3R`](https://docs.nvidia.com/cuda/archive/10.2/npp/group__image__packed__to__planar__channel__copy.html#gac5f584f230216c6b62e9c5b7e4b7a18b) for divide channel plane


Since nppi operations require aligned memory regions (i.e., it may include a padding area), you may need to handle such a padding area properly when copying the actual data onto the network input